stats/cloud/logs.go

```diff
@@ -0,0 +1,212 @@
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package cloud
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/mailru/easyjson"
+	"github.com/sirupsen/logrus"
+)
+
+//easyjson:json
+type msg struct {
+	Streams        []msgStreams        `json:"streams"`
+	DroppedEntries []msgDroppedEntries `json:"dropped_entries"`
+}
+
+//easyjson:json
+type msgStreams struct {
+	Stream map[string]string `json:"stream"`
+	Values [][2]string       `json:"values"` // this can be optimized
+}
+
+//easyjson:json
+type msgDroppedEntries struct {
+	Labels    map[string]string `json:"labels"`
+	Timestamp string            `json:"timestamp"`
+}
+
+func getLevelsStr(level string) ([]string, error) {
+	lvl, err := logrus.ParseLevel(level)
+	if err != nil {
+		return nil, fmt.Errorf("unknown log level %s", level) // specifically use a custom error
+	}
+	index := sort.Search(len(logrus.AllLevels), func(i int) bool {
+		return logrus.AllLevels[i] > lvl
+	})
+	result := make([]string, index)
+	for i, lvl := range logrus.AllLevels[:index] {
+		result[i] = lvl.String()
+	}
+
+	return result, nil
+}
+
+func (m *msg) Log(logger logrus.FieldLogger) {
+	var level string
+
+	for _, stream := range m.Streams {
+		fields := labelsToLogrusFields(stream.Stream)
+		var ok bool
+		if level, ok = stream.Stream["level"]; ok {
+			delete(fields, "level")
+		}
+
+		for _, value := range stream.Values {
+			nsec, _ := strconv.Atoi(value[0])
+			e := logger.WithFields(fields).WithTime(time.Unix(0, int64(nsec)))
+			lvl, err := logrus.ParseLevel(level)
+			if err != nil {
+				e.Info(value[1])
+				e.Warn("last message had unknown level " + level)
+			} else {
+				e.Log(lvl, value[1])
+			}
+		}
+	}
+
+	for _, dropped := range m.DroppedEntries {
+		nsec, _ := strconv.Atoi(dropped.Timestamp)
+		logger.WithFields(labelsToLogrusFields(dropped.Labels)).WithTime(time.Unix(0, int64(nsec))).Warn("dropped")
+	}
+}
+
+func labelsToLogrusFields(labels map[string]string) logrus.Fields {
+	fields := make(logrus.Fields, len(labels))
+
+	for key, val := range labels {
+		fields[key] = val
+	}
+
+	return fields
+}
+
+func parseFilters(id, level string) ([]string, error) {
+	idFilter := `test_run_id="` + id + `"`
+
+	lvls, err := getLevelsStr(level)
+	if err != nil {
+		return nil, err
+	}
+	levelFilter := `level=~"(` + strings.Join(lvls, "|") + `)"`
+
+	return []string{idFilter, levelFilter}, nil
+}
+
+func (c *Config) getRequest(referenceID string, start time.Duration, limit int) (*url.URL, error) {
+	u, err := url.Parse(c.LogsHost.String)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't parse cloud logs host %w", err)
+	}
+
+	filters, err := parseFilters(referenceID, logrus.InfoLevel.String())
```

Comment:
Why are you dealing with strings? Why not just use `logrus.Info`? 

And, more importantly, why are you actually restricting the level to `info` here? I don't think k6 needs to specify a level when it's tailing logs, given that they are limited in the cloud, it should just get whatever was emitted by the cloud. Or, alternatively, this should behave like `k6 run` and show `info` level normally and `debug` if `--verbose` is enabled...

Reply:
carried it over from the k6cloudlogs, dropped it now